File: pysaurus/interface/web/src/components/Pagination.js
import { FormPaginationGoTo } from "../forms/FormPaginationGoTo.js";
import { DialogSearch } from "../dialogs/DialogSearch.js";
import { capitalizeFirstLetter } from "../utils/functions.js";

export class Pagination extends React.Component {
	constructor(props) {
		// singular: str
		// plural: str
		// nbPages: int
		// pageNumber: int
		// onChange: function(int)
		// onSearch? function(str)
		super(props);
		this.onFirst = this.onFirst.bind(this);
		this.onNext = this.onNext.bind(this);
		this.onLast = this.onLast.bind(this);
		this.onPrevious = this.onPrevious.bind(this);
		this.go = this.go.bind(this);
		this.look = this.look.bind(this);
	}

	render() {
		const singular = this.props.singular;
		const plural = this.props.plural;
		const nbPages = this.props.nbPages;
		const pageNumber = this.props.pageNumber;
		return nbPages ? (
			<span className="navigation py-1 text-center">
				<button
					className="first"
					disabled={pageNumber === 0}
					onClick={this.onFirst}>
					&lt;&lt;
				</button>
				<button
					className="previous"
					disabled={pageNumber === 0}
					onClick={this.onPrevious}>
					&lt;
				</button>
				<span
					{...(this.props.onSearch
						? { className: "go", onClick: this.look }
						: {})}>
					{capitalizeFirstLetter(singular)}
				</span>
				<span className="go clickable" onClick={this.go}>
					{pageNumber + 1}/{nbPages}
				</span>
				<button
					className="next"
					disabled={pageNumber === nbPages - 1}
					onClick={this.onNext}>
					&gt;
				</button>
				<button
					className="last"
					disabled={pageNumber === nbPages - 1}
					onClick={this.onLast}>
					&gt;&gt;
				</button>
			</span>
		) : (
			<span className="navigation py-1 text-center">
				<em>0 {plural}</em>
			</span>
		);
	}

	onFirst() {
		if (this.props.pageNumber !== 0) {
			this.props.onChange(0);
		}
	}

	onPrevious() {
		if (this.props.pageNumber > 0) {
			this.props.onChange(this.props.pageNumber - 1);
		}
	}

	onNext() {
		if (this.props.pageNumber < this.props.nbPages - 1) {
			this.props.onChange(this.props.pageNumber + 1);
		}
	}

	onLast() {
		if (this.props.pageNumber !== this.props.nbPages - 1) {
			this.props.onChange(this.props.nbPages - 1);
		}
	}

	go() {
		Fancybox.load(
			<FormPaginationGoTo
				nbPages={this.props.nbPages}
				pageNumber={this.props.pageNumber}
				onClose={(pageNumber) => {
					if (pageNumber !== this.props.pageNumber)
						this.props.onChange(pageNumber);
				}}
			/>
		);
	}

	look() {
		Fancybox.load(
			<DialogSearch title={"Search first:"} onSearch={this.props.onSearch} />
		);
	}
}

File: pysaurus/interface/using_videre/pagination.py
from typing import Callable

import videre


class Pagination(videre.Row):
    __wprops__ = {}
    __slots__ = ("_on_change", "_page_number", "_nb_pages")

    def __init__(
        self, nb_pages: int, page_number: int, on_change: Callable[[int], None]
    ):
        self._on_change = on_change
        self._page_number = page_number
        self._nb_pages = nb_pages
        super().__init__(
            [
                videre.Button("<<", on_click=self._on_first, disabled=page_number == 0),
                videre.Button(
                    "<", on_click=self._on_previous, disabled=page_number == 0
                ),
                videre.Text(f"Page {page_number + 1}/{nb_pages}"),
                videre.Button(
                    ">", on_click=self._on_next, disabled=page_number == nb_pages - 1
                ),
                videre.Button(
                    ">>", on_click=self._on_last, disabled=page_number == nb_pages - 1
                ),
            ],
            vertical_alignment=videre.Alignment.CENTER,
            space=10,
        )

    def _on_first(self, *args):
        if self._page_number != 0:
            self._on_change(0)

    def _on_last(self, *args):
        if self._page_number != self._nb_pages - 1:
            self._on_change(self._nb_pages - 1)

    def _on_previous(self, *args):
        if self._page_number > 0:
            self._on_change(self._page_number - 1)

    def _on_next(self, *args):
        if self._page_number < self._nb_pages - 1:
            self._on_change(self._page_number + 1)
